presets/ragengine/main.py

```diff
@@ -253,6 +253,83 @@ async def list_documents_in_index(
     except Exception as e:
         raise HTTPException(status_code=500, detail=str(e))
 
+@app.post(
+    "/indexes/{index_name}/documents/{doc_id}",
+    response_model=DocumentResponse,
+    summary="Update document in an Index by its ID",
+    description="""
+    Update document in an Index by its ID.
+
+    ## Request Example:
+    ```json
+    POST /indexes/test_index/documents/test_doc_id
+    {"text": "Sample document text.", "metadata": {"author": "John Doe", "category": "example"}}
+    ```
+
+    ## Response Example:
+    ```json
+    {
+        "doc_id": "123456",
+        "text": "Sample document text.",
+        "metadata": {"author": "John Doe"},
+        "is_truncated": true
+    },
+    ```
+    """,
+)
+async def update_document_in_index(
+    index_name: str,
+    doc_id: str,
+    request: UpdateDocumentRequest,
+):
+    try:
+        decoded_index_name = unquote(index_name)
+        decoded_doc_id = unquote(doc_id)
+        await rag_ops.update_document(
+            index_name=decoded_index_name,
+            doc_id=decoded_doc_id,
+            new_doc=request
+        )
+
+        return request
+    except Exception as e:
+        raise HTTPException(status_code=500, detail=str(e))
+
+@app.delete(
+    "/indexes/{index_name}/documents/{doc_id}",
+    summary="Delete Document in an Index by its ID",
+    description="""
+    Delete Document in an Index by its ID.
+
+    ## Request Example:
+    ```
+    DELETE /indexes/test_index/documents/test_doc_id
+    ```
+
+    ## Response Example:
+    ```json
+    {
+      "message": "Successfully deleted document test_doc_id from index test_index."
+    }
+    ```
+    """,
+)
+async def delete_document_in_index(
```

Comment:
do we want update/delete to fail on non existent document ids?

Reply:
this happens. we dont want to fail entire request, we return something like the ones we could delete and/or print error message

{
  "deleted": ["doc1", "doc2"],
  "not_found": ["doc3"]
}

Reply:
I think we can just pass whatever err msg returned by the storage server if it implements the delete api.

Reply:
for update I went with {"updated_documents":[], "skipped_documents":[]} but in reality the api handles creating documents if they dont exist in the update call so we likely wont see skipped documents

Reply:
updated the api to what we discussed!